### PR TITLE
Adds HMAC Validator

### LINKED ISSUE #35: HMAC Validator

### PR DESCRIPTION
**Pull request by KadoBOT**
resolves #35

### CODE REVIEW

File: src/hmacvalidator/hmacvalidator.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/adyen/adyen-go-api-library/src/notification"
+)
+
+func CalculateHmac(data interface{}, secret string) (string, error) {
+	switch val := data.(type) {
```
Comment (deepu105) [inline src/hmacvalidator/hmacvalidator.go:17]: I updated this to reduce type casts

File: src/hmacvalidator/hmacvalidator.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	expectedSign, err := CalculateHmac(notificationRequestItem, key)
+	if err != nil {
+		return false
+	}
+	merchantSign := (*notificationRequestItem.AdditionalData)["HmacSignature"]
+	return expectedSign == merchantSign
+}
+
+func GetDataToSign(notificationRequestItem interface{}) string {
+	switch item := notificationRequestItem.(type) {
```
Comment (deepu105) [inline src/hmacvalidator/hmacvalidator.go:36]: simplified code here

File: src/hmacvalidator/hmacvalidator.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/adyen/adyen-go-api-library/src/notification"
+)
+
+func CalculateHmac(data interface{}, secret string) (string, error) {
```
Comment (deepu105) [inline src/hmacvalidator/hmacvalidator.go:16]: can you add a method comment saying what this does

File: src/hmacvalidator/hmacvalidator.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	switch val := data.(type) {
+	case string:
+		return encode(val, secret)
+	default:
+		src := GetDataToSign(data)
+		return encode(src, secret)
+	}
+}
+
+func ValidateHmac(notificationRequestItem notification.NotificationRequestItem, key string) bool {
```
Comment (deepu105) [inline src/hmacvalidator/hmacvalidator.go:26]: can you add a method comment saying what this does

**Comment by deepu105** [comment]
something weird is going on with the tests

File: src/hmacvalidator/hmacvalidator.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	case map[string]string:
+		keys := make([]string, 0)
+		values := make([]string, 0)
+
+		for k, v := range item {
+			keys = append(keys, replacer(k))
+			values = append(values, replacer(v))
+		}
+
+		return strings.Join(keys, ":") + ":" + strings.Join(values, ":")
```
Comment (deepu105) [inline src/hmacvalidator/hmacvalidator.go:58]: oh most important thing, maps in go are not sorted and you cannot sort them, means you cannot rely on their order for anything, so this won't work. What kind of order is expected in output, maybe you can sort the key and value array

File: src/hmacvalidator/hmacvalidator_test.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	MerchantReference:   "reference",
+	OriginalReference:   "originalReference",
+	PaymentMethod:       "VISA",
+	PspReference:        "pspReference",
+	Reason:              "reason",
+	Success:             "true",
+}
+
+func Test_Hmacvalidator(t *testing.T) {
+	t.Run("Get correct data", func(t *testing.T) {
```
Comment (deepu105) [inline src/hmacvalidator/hmacvalidator_test.go:33]: can we group tests by method?

File: src/hmacvalidator/hmacvalidator.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
+package hmacvalidator
```
Comment (rikterbeek) [inline src/hmacvalidator/hmacvalidator.go:1]: shall we call it utils/hmacvalidator.go ?

Comment (deepu105) [reply src/hmacvalidator/hmacvalidator.go:1]: you can, but package will still be `hmacvalidator` only the import path will change

**Pull request merged**